src/PowerShellEditorServices/Language/FoldingReference.cs

```diff
@@ -60,4 +61,33 @@ public class FoldingReference: IComparable<FoldingReference>
             return string.Compare(this.Kind, that.Kind);
         }
     }
+
+    /// <summary>
+    /// A class that holds a list of FoldingReferences and ensures that when adding a reference that the
+    /// folding rules are obeyed, e.g. Only one fold per start line
+    /// </summary>
+    public class FoldingReferenceList : Dictionary<int, FoldingReference>
+    {
+        /// <summary>
+        /// Adds a FoldingReference to the list and enforces ordering rules e.g. Only one fold per start line
+        /// </summary>
+        public void SafeAdd(FoldingReference item)
+        {
+            if (item == null) { return; }
+            FoldingReference currentItem;
+            TryGetValue(item.StartLine, out currentItem);
```

Comment:
Doesn't the version of C# we use support `TryGetValue(item.StartLine, out FoldingReference currentItem)`?  Also, it is a bit unusual to see a `Try*` method, which returns a bool, not used inside an `if` condition e.g. `if (TryGetValue(...)) { ... }`.  It appears you are relying on the TryGetValue to set the current item to null but you could just as well use the return value from `TryGetValue` to determine if `currentItem` will be valid or not.  IMO the intent would be more clear in the code.

Reply:
Apologies.  My rubyisms coming through again.  Fixed to use more C#-isms